### PR TITLE
stream info: add attempt count for HTTP requests and TCP proxy

### PR DESCRIPTION
**Pull request by junr03**
Commit Message: stream info - add attempt count for HTTP requests and TCP proxy
Risk Level: low
Testing: UT

### CODE REVIEW

**Comment by junr03** [comment]
cc @goaway

**Comment by goaway** [comment]
Thanks @junr03!

File: source/common/router/router.cc [hunk 1]
```diff
@@ -602,6 +602,7 @@ Http::FilterHeadersStatus Filter::decodeHeaders(Http::RequestHeaderMap& headers,
       headers.setEnvoyAttemptCount(attempt_count_);
     };
   }
+  callbacks_->streamInfo().setAttemptCount(attempt_count_);
```
Comment (snowp) [inline source/common/router/router.cc:605]: I'm not sure if this is the right place, isn't this just called once per request prior to retries happening? I think we'd need to set this somewhere in the retry logic?

Comment (junr03) [reply source/common/router/router.cc:605]: @snowp I am also setting in doRetry(): https://github.com/envoyproxy/envoy/pull/17583/files#diff-e3584a650478334f7373c0df344c85d499b5123f5237ce2bbabaad5ee53b899eR1609

**Review by ggreenway — COMMENTED** [review]
This would be trivial to add support for in tcp_proxy. Any chance you could do that to have less fragmentation in supported features?

**Comment by junr03** [comment]
@ggreenway fair point, update to include in tcp proxy.

File: source/common/router/router.cc (hunk 1 above)
Comment (snowp) [reply source/common/router/router.cc:605]: Not sure how I missed this, that seems right to me!

File: envoy/stream_info/stream_info.h [hunk 2]
```diff
@@ -593,6 +593,17 @@ class StreamInfo {
    * @return Network filter chain name of the downstream connection.
    */
   virtual const std::string& filterChainName() const PURE;
+
+  /**
+   * @param attempt_count, the number of times the request was attempted upstream.
+   */
+  virtual void setAttemptCount(uint32_t attempt_count) PURE;
```
Comment (ggreenway) [inline envoy/stream_info/stream_info.h:600]: Thought for a future day: do we want to split StreamInfo into downstream and upstream? Some fields are only applicable to one or the other; attempt count only makes sense on the downstream side.

**Review by ggreenway — APPROVED** [review]
LGTM. Thanks for adding tcp_proxy code.

File: envoy/stream_info/stream_info.h (hunk 2 above)
Comment (junr03) [reply envoy/stream_info/stream_info.h:600]: I had similar thoughts while having a [discussion](https://github.com/envoyproxy/envoy/pull/17512#discussion_r681610376) with @soulxu on a similar PR to this one. 

Created https://github.com/envoyproxy/envoy/issues/17594

**Pull request merged**